File: 2-1/2-1-Java/MrCoxallStack.java
/*
 * This is a program that prints out stuff.
 * @author  Jakub Malhotra
 * @version 1.0
 * @since   2024-10-16
 */

import java.util.ArrayList;
import java.util.List;

/**
 * This is a standard program.
 */
public class MrCoxallStack {

    /**
     * Represents a stack data structure implemented using a List.
     * This list stores elements as strings and is public.
     */
    public List<String> stackAsList = new ArrayList<>();

    /**
     * Pushes the specified string onto the top of the stack.
     *
     * @param pushedString the string to be pushed onto the stack
     */
    public void pushItem(final String pushedString) {
        // Push the value onto the stack
        this.stackAsList.add(pushedString);
    }

    /**
     * Returns a string representation of the stack.
     *
     * @return the string representation of the stack
     */
    public String showStack() {
        // Return the stack
        final StringBuilder stackValues = new StringBuilder();

        for (String value : this.stackAsList) {
            stackValues.append(value).append(",");
        }

        if (stackValues.length() > 0) {
            // Remove the trailing comma
            stackValues.deleteCharAt(stackValues.length() - 1);
        }

        return stackValues.toString();
    }
}

File: 2-3/2-3-TS/mrCoxallStack.ts
/**
 * Class representing a stack for storing strings.
 * @author  Jakub Malhotra
 * @version 2.1 
 * @since   2024-10-21
 */
export default class MrCoxallStack {
    private stackAsList: string[];

    /**
     * Create a new stack.
     */
    constructor() {
        this.stackAsList = [];
    }

    /**
     * Getter to return the current stack as a string with comma-separated values.
     * @returns {string} The stack as a string.
     */
    get showStack(): string {
        if (this.isEmpty) {
            return "The stack is empty!";
        }
        
        return this.stackAsList.join(", ");
    }

    /**
     * Getter to check if the stack is empty.
     * @returns {boolean} True if the stack is empty, false otherwise.
     */
    isEmpty(): boolean {
        if (this.stackAsList.length === 0) {
            // Stack is empty
            return true
        }
        // Stack is not empty
        return false;
    }    

    /**
     * Getter to get the size of the stack.
     * @returns {number} The number of elements in the stack.
     */
    get size(): number {
        return this.stackAsList.length;
    }

    /**
     * Push a string onto the stack.
     * @param {string} pushedString - The string to be pushed onto the stack.
     */
    pushItem(pushedString: string): void {
        this.stackAsList.push(pushedString);
    }

    /**
     * Pop a string from the stack.
     * @returns {string} The popped string, or a message if the stack is empty.
     */
    popItem(): string {
        if (this.isEmpty) {
            return "The stack is empty!";
        }
        
        const poppedValue = this.stackAsList.pop();
        
        // Handle the case where poppedValue might be undefined
        if (poppedValue === undefined) {
            return "The stack is empty!";
        }
        
        return poppedValue;
    }
    
}
